mavlink: Allow setting the frequency requested for individual messages

rafaellehmkuhl opened the pull request:
https://github.com/user-attachments/assets/95d3d2f6-e855-4f7a-b15e-1096f06bf85f


To be merged after #1912 
Fix #1857

ES-Alexander commented:
Cool feature to have!

Some concerns:
1. This adds an additional page for MAVLink stuff
    - Could it instead be a second expandable section within the existing MAVLink inspector? That way you configure data requests in the same place where you can directly see that it's coming in
2. From a quick look through the code, this doesn't seem to account for [stream groups](https://ardupilot.org/sub/docs/ArduSub_MAVLink_Messages.html#stream-groups)
    - I'm not sure how best to reconcile the interface for group vs individual requests, although I suppose a minimal viable solution would be to _read_ the group rates and display those as defaults for the relevant messages, but then allow overriding those with individual requests (assuming that's how the autopilot firmware handles that - should probably be checked)
    - It would be _nice_ if there can be some interface component showing that a message rate is currently being governed by a group request, and possibly having another section to configure groups (or even listing the messages by group, with an "ungrouped" group to handle the extras), and then having a group slider with the name, plus individual sliders for overrides, e.g.
        <img width="677" alt="image" src="https://github.com/user-attachments/assets/8857cb36-5b56-4649-a8e8-a69f0a75ca15" />

patrickelectric commented on src/stores/mainVehicle.ts:
```diff
@@ -603,6 +611,37 @@ export const useMainVehicleStore = defineStore('main-vehicle', () => {
+    customMessageIntervals.value.data[messageType as keyof typeof defaultMessageFrequency] = interval
+
+    // Schedule the actual update to the vehicle
+    await (mainVehicle.value as ArduPilot).setMessageInterval(messageType as MAVLinkType, interval)
```
This should only be set if the current frequency of the message is lower than the expected frequency. Other systems may set the different message frequencies that could be higher than cockpit, and cockpit should not make it lower.

rmackay9 replied:
I don't fully understand the code or comments but I'd like to clarify that we sometimes want to use the message interval feature to reduce the rate at which a message is sent because the current rate is consuming too much bandwidth.  We sometimes want to stop a message from being sent.  Also the user is the one changing the controls so I think they should be obeyed more than other parts of the system.  We do sometimes see other parts of the system (normally MAVProxy or Mission Planner) overwriting the rates and the solution is normally to stop them from doing this.  For example MP has a screen to set the desired rates and it accepts -1 which means to leave them alone and not try to set them at all.  MAVProxy also has a way to stop it from setting the desired rates but it's a startup command only I believe.  MAVProxy is quite annoying in this way actually

patrickelectric replied:
I agree, but this frequencies are not just for telemetry but for inner usage of the MAVLink data as well inside BlueOS. I _hope_ the user knows what he's doing, this is a feature that could be under a _**pirate mode**_ since things can go wrong.

rafaellehmkuhl replied:
@patrickelectric I think it's fair to put it under a pirate mode. Will add it.

@rmackay9 I will add the possibility of setting explicitly for cockpit *not* to set the interval as well.

rmackay9 commented:
I tried to test this by setting the "BlueOS Version" page's "Remote Versions" to "rafaellehmkuhl/cockpit-public" but that didn't work for some reason.  The issue is surely that I don't yet know BlueOS well enough.

ES-Alexander commented:
> I tried to test this by setting the "BlueOS Version" page's "Remote Versions" to "rafaellehmkuhl/cockpit-public" but that didn't work for some reason.

@rmackay9 The BlueOS Version page is for managing your version of BlueOS. To install Extensions you'll want to use the [Extensions manager](https://blueos.cloud/docs/latest/usage/advanced/#extensions-manager). In the installed tab you can click the plus circle in the bottom right corner, then fill out the relevant details for the custom extension you want to install :-)

In this case my setup is:

| Field | Value | Notes |
---|---|---
Extension Identifier | `rafaellehmkuhl.cockpit` | user choice, just pick something reasonable and unique - I usually go for the image name with a dot instead of the slash
Extension Name | `Cockpit Rafael` | user choice, just pick something descriptive enough
Docker image | `rafaellehmkuhl/cockpit` | this needs to be correct, and can be checked [in Docker Hub](https://hub.docker.com/u/rafaellehmkuhl) if necessary
Docker tag | `allow-individual-message-interval` | branch name - I usually copy from the GitHub PR header, and just delete the username part.
Original Settings | ...[^1] | copy from an existing install of the extension, or get them from the `permissions` field in the repo Dockerfile, removing the string-escaping backslashes

Once you've got it set up you can disable any other installs of the same extension (to avoid confusion in the sidebar), and then to review future pull requests from the same author you can Edit the extension (requires [Pirate Mode](https://blueos.cloud/docs/latest/usage/advanced/#pirate-mode)) and just update the Docker tag field :-)

[^1]: ```
    {
      "ExposedPorts": {
        "8000/tcp": {}
      },
      "HostConfig": {
        "PortBindings": {
          "8000/tcp": [
            {
              "HostPort": ""
            }
          ]
        }
      }
    }
    ```

rmackay9 commented on src/libs/utils.ts:
```diff
@@ -275,3 +275,12 @@ export const getUnindentedString = (str: string): string => {
+ * @returns {number} The interval in microseconds
+ */
+export const frequencyHzToIntervalUs = (frequencyHz: number): number => {
+  return 1000000 / frequencyHz
```
we should be careful of divide by zero here perhaps

rafaellehmkuhl replied:
Yup.
Will perform the changes to allow the disabling as well.

rafaellehmkuhl replied:
@rmackay9 curiosity: when the docs specify [a default rate](https://mavlink.io/en/messages/common.html#MAV_CMD_SET_MESSAGE_INTERVAL), this means a rate implemented on the vehicle firmware or it means to ignore whatever was set by the sender and consider what was sent by other systems?

ES-Alexander replied:
MAVLink doesn't specify (i.e. it would be up to the receiving component to decide what its default rate is).

If it's relevant, [ArduPilot's implementation](https://github.com/ArduPilot/ardupilot/blob/593ea27ce52b7fea2372c91321dea39d411739c4/libraries/GCS_MAVLink/GCS_Common.cpp#L6955) uses either:
1. fixed defaults for non-streamed messages like `HEARTBEAT`s,
2. the parameter value that is currently configured for the stream group which contains the message in question, or
3. [no sending](https://github.com/ArduPilot/ardupilot/blob/593ea27ce52b7fea2372c91321dea39d411739c4/libraries/GCS_MAVLink/GCS_Common.cpp#L3153), if the message is not included in a stream group (the "Ungrouped" section of messages in my example image)

While the user hasn't specified a value _in Cockpit_ / Cockpit doesn't have a reason for requesting a particular rate of that message, then it can either 
1. avoid requesting it (to receive whatever rate has been requested using other software, defaulting to the autopilot's stream rates)
2. explicitly request it at its default rate from the autopilot (at risk of overriding other software configurations, if they've requested individual message rates instead of configuring the stream groups).

rmackay9 replied:
Hi @ES-Alexander,

Re the last part of the last comment above, I think if Cockpit doesn't have a reason for requesting a particular rate for a message it should avoid requesting it.  Cockpit should just leave it as it is, maybe that means leaving a field blank in this new extension.

rmackay9 replied:
@rafaellehmkuhl, sorry to take so long to reply to your question.  I think if Cockpit sends a message-interval to the autopilot requesting the default rate then the autopilot will probably revert that message to what the stream-rate settings have it as.  the stream rates are initialised by the SRx_ parameters (recently renamed to MAVx_).  I don't think that cockpit should try to be too smart about this, I think it should just allow the user to set the rate for any message and leave it at that, don't try and check if the request has actually been achieved.

I think what would be helpful though is if the user could see what rate the messages are currently arriving at approximately (e.g. what's shown in the mavlink inspector) and also the currently sent messages should appear at the top.  MAVLink has a lot of messages but AP doesn't use them all and the user shouldn't really have to sift through all of them very often

rmackay9 commented on src/stores/appInterface.ts:
```diff
@@ -19,18 +19,18 @@ export enum SubMenuName {
-  SettingsDataLake = 'settings-datalake',
-  ToolsMAVLink = 'tools-mavlink',
-  ToolsDataLake = 'tools-datalake',
+  SettingsGeneral = 'SettingsGeneral',
```
I'm not qualified to review this code of course but I'm surprised that we're changing the formatting of all these lines.  I'd expect one new entry but I wouldn't have expected existing entries to change

patrickelectric replied:
Style changes should at least break in another commit

rmackay9 commented:
Hi @rafaellehmkuhl,

Another small thing I noticed, it would probably be best if the mavlink messages appeared in alphabetical order

rafaellehmkuhl commented:
@ES-Alexander @rmackay9 @patrickelectric I have refactored the functionality, as can be seen below.

The four options now are:
- Disable (don't send that message)
- Don't touch (Cockpit don't try to define the interval, even if its on its defaults)
- Vehicle default (whatever the vehicle decided)
- Custom (set a specific frequency)

The user options are stored in sync with the vehicle and automatically aplied on Cockpit's boot or when the option is changed.

I've also added a forms under the table so the user can specify the interval for any option that is not there (the table is populated with Cockpit's defaults).

This MAVLink configuration menu only appears if the user has pirate-mode activated (in the general menu).

https://github.com/user-attachments/assets/97c34e0e-6d44-4910-bd58-60550eaf7f72

rmackay9 commented:
thanks very much for this, sorry for taking a while to test, I hope to get to this tomorrow but 100% by (my) monday

rafaellehmkuhl commented:
> @rafaellehmkuhl,
> 
> 
> 
> I installed this (I think) but for some reason I don't see the new MAVlink interval feature.  I think it should appear in Cockpit under Settings and then just below "Actions"
> 
> 
> I completely uninstalled the default Cockpit and installed the new one and below we can see it is running and then settings I applied.  I'm also using Pirate mode
> 
> 
> 
> Any idea what I might be doing incorrectly?
> 
> 
> 
> By the way, I wonder if it might be better to somehow put this under the existing Telemetry setting?  .. or maybe "Telemetry" is really "Telemetry OSD" so it's not the same?

I forgot to mention @rmackay9! You have to go to the general settings and click to enable pirate mode!

rmackay9 commented:
Hi @rafaellehmkuhl,

Excuse my ignorance but I guess this is a cockpit specific pirate mode?  I've already put BlueOS into Pirate mode.

I've gone into Cockpit, Settings, General, "Manage Cockpit Settings" and I don't see any setting containing "pirate".  Maybe this is because https://github.com/bluerobotics/cockpit/pull/1912 is not included in the branch?

rafaellehmkuhl commented:
> Hi @rafaellehmkuhl,
> 
> Excuse my ignorance but I guess this is a cockpit specific pirate mode? I've already put BlueOS into Pirate mode.
> 
> I've gone into Cockpit, Settings, General, "Manage Cockpit Settings" and I don't see any setting containing "pirate". Maybe this is because #1912 is not included in the extension yet?
> 
> Thanks again for all your efforts on this!

Don't worry Randy haha

This is how you enable it (:

https://github.com/user-attachments/assets/bcdccc05-2d52-43e1-86d1-050b293b905f

rmackay9 reviewed (approved):
I'm not qualified to review the code but I've tested the new feature and it seems to work, thanks very much!

rafaellehmkuhl commented:
> I'm not qualified to review the code but I've tested the new feature and it seems to work, thanks very much!

Thank you for the feedback Randy! I will try to include it in the next Beta release.

ArturoManzoli commented:
Code looks fine to me;

My only change would be to redistribute the elements horizontally on the table, so no void space would be left on the right part of the screen

![image](https://github.com/user-attachments/assets/e50bad2d-1668-46bc-8bee-b4b465fbc165)

rafaellehmkuhl commented:
> Code looks fine to me;
> 
> My only change would be to redistribute the elements horizontally on the table, so no void space would be on the right part of the screen

Thanks for the tip!
I ended up changing it a lot. It's now a lot more compact and can fit even when in half-screen:

<img width="833" alt="image" src="https://github.com/user-attachments/assets/c41849b3-0658-4884-82b8-b52f48220d11" />

ArturoManzoli reviewed (approved):
Looking very good now

rmackay9 commented:
Yay!  thanks for this!!

rafaellehmkuhl commented:
> Yay!  thanks for this!!

No problem Randy!
It was included in beta4, released today.
Thanks again for the idea and feedback.
